homeassistant/components/mqtt/binary_sensor.py

```diff
@@ -231,3 +282,11 @@ def force_update(self):
     def unique_id(self):
         """Return a unique ID."""
         return self._unique_id
+
+    @property
+    def state(self):
+        """Return the state of the binary sensor."""
+        if self._available:
```

Comment:
Same comment as above.

Reply:
This is because this sensor is inheriting the `state` property from the `binary_sensor.py`.

https://github.com/home-assistant/home-assistant/blob/6acfede512dcb2c0f41b5d591a3a2794e4299192/homeassistant/components/binary_sensor/__init__.py#L153

The state in `binary_sensor.py` only returns either `STATE_ON` and `STATE_ON`.  I implemented it so that it supports the 3 states, `STATE_ON`, `STATE_OFF`, and `STATE_UNAVAILABLE`.

Is there another way to do this, by making use from a state property else where?

Reply:
We're not allowed to overwrite the state property of the base binary_sensor entity class. We need to keep the entity interface the same.

The availability is set separately from the state property.

Reply:
Hey Martin, thanks for getting back to me. Will you please help me solve this?

What should we do instead of this?  I came up with this because that’s how it is handled by the normal MQTT sensor, the tests all check for STATE_UNAVILABLE. 

Thanks for helping out!

Reply:
If the entity updates `available` property and returns `False` there, the entity base class will set state to `STATE_UNAVAILABLE`. This is the interface we need to use.

Reply:
Hi @MartinHjelmare, @emontnemery,

I do set `self._available = False` in my expiration trigger, 

https://github.com/home-assistant/home-assistant/blob/10762e72c681a753aec65ccf648e2a92395d1d46/homeassistant/components/mqtt/binary_sensor.py#L245

but the `state` property on https://github.com/home-assistant/home-assistant/blob/6acfede512dcb2c0f41b5d591a3a2794e4299192/homeassistant/components/binary_sensor/__init__.py#L153 only ever returns `STATE_ON` or `STATE_OFF`.

I just tested this by removing my `state` property, so that the one in `binary_sensor/__init__.py` gets called.  The sensor then never returns `STATE_UNAVILABLE`.

Is there something else I need to do for this to work?

Reply:
Have you set an availability topic in the config?

Reply:
I haven't specifically no, because like on the MQTT sensor, the expiry works without having to specify an availability topic.  

Are you saying that if we want this functionality to work, we will always need to specify an availability topic in the config?

Reply:
Currently that's how the availability mixin for mqtt works.

https://github.com/home-assistant/home-assistant/blob/3714cdaa6dc0475b76f50a1aa6c4fab237225f32/homeassistant/components/mqtt/__init__.py#L1132-L1135

Reply:
That means that the tests will only pass if ALL implementations now get the availability topic set in the config.  If you don't set it, you can't set the component to STATE_UNAVAILABLE.  

If we update all the tests to set an availability topic, everyone that has existing config without availability topics will break.

I still can't see how we can make it work without having a state property that outputs 3 possible values (`STATE_ON`, `STATE_OFF`, and `STATE_UNAVAILABLE`).

Maybe I'm missing something, so if you don't mind, please give me a hand so that we can get this item over the line :)

Thanks,

Reply:
Why can't we leave it optional?

Reply:
If you leave the state topic option out, the state can never be `STATE_UNAVAILABLE`.  Which means the tests fail if you have the `expire_after` set.  

So what I mean by that is, even if you set an `expire_after`, the state can not possibly be `STATE_UNAVAILABLE` when it should.

Reply:
The user could choose by setting the availability topic if the state should expire to off or to unavailable. Maybe that will be confusing?

Reply:
I think we should simply set the binary sensor to unavailable and not mess with the state.

Looking at your problem description:
_Each of these sensors have a heartbeat (I believe it's every 51 minutes) which sends a 'I'm here' back to the mqtt (and thus Home Assistant).

My problem was that if I lose the link between CC2530/ESP8266 and zigbee2mqtt, Home Assistant keeps displaying the same state that it has been, e.g. no motion, or door closed._

This fits well with setting the state to unvailable.

Actually, why doesn't zigbee2mqtt just do that?

Reply:
Hi @emontnemery, thanks for getting back to me.

How do I set the `binary_sensor` to unavailable?  It never returns 'unavailable', because it only returns `STATE_ON` or `STATE_OFF`.  Are you saying if I set the availability on the sensor to `false` it will show as unavailable in the front end, even if the state still returns `STATE_ON`?

zigbee2mqtt only has an availability topic for the bridge itself, so we already use that.  If zigbee2mqtt goes offline, all your devices on that 'bridge' is marked as unavailable.  This doens't handle the case that my PR is handling.  I'm handling if only an individual device doesn't ping back in the allotted time.  

I guess zigbee2mqtt could build in availability topics per sensor, and handle those timings and then send that availability across.  I just genuinely thought that this was a good idea, to keep the consistency with the mqtt sensor, because it does exactly what we want already.

Cheers,